src/operator/subgraph/subgraph_op.h

```diff
@@ -0,0 +1,181 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+#ifndef MXNET_OPERATOR_SUBGRAPH_SUBGRAPH_OP_H_
+
+#include <nnvm/graph.h>
+#include <nnvm/pass.h>
+
+namespace mxnet {
+
+namespace op {
+
+namespace sg {  // sg stands for subgraph
+
+struct SimpleNode;
+using SimpleNodePtr = std::shared_ptr<SimpleNode>;
+
+struct SimpleNode {
+  static SimpleNodePtr Create() {
+    return std::make_shared<SimpleNode>();
+  }
+  SimpleNode() : label(-1), node(nullptr) {}
+  int label;
+  nnvm::Node* node;
+  // key is node ptr
+  // value is the index array standing for the entry indices
+  // in key->inputs that use this->node as input node
+  std::unordered_map<nnvm::Node*, std::vector<int>> outputs;
+};
+
+}
+
+/*
+ * This provides criteria for selecting nodes in a subgraph.
+ * When a node is passed to this object, the selection criteria may be changed.
+ * We can also specify what links we should use when traversing the neighbor
+ * nodes.
+ */
+class SubgraphSelect {
+ public:
+  virtual ~SubgraphSelect() {
+  }
+  /*
+   * Given a set of nodes that have been selected so far for a subgraph, determine
+   * if the input node should be selected for a subgraph.
+   */
+  virtual bool Select(const nnvm::Node &n,
+                      const std::vector<sg::SimpleNode*> *subgraph_nodes) = 0;
+  virtual bool UseIncomingLink() const = 0;
+  virtual bool UseOutgoingLink() const = 0;
+  virtual void Reset() {
+  }
+};
+
+using SubgraphSelectPtr = std::shared_ptr<SubgraphSelect>;
+
+/*
+ * This is the interface of the subgraph operator that executes the computation
+ * in the subgraph.
+ */
+class SubgraphOperator {
+  nnvm::Symbol subgraph_sym_;
```

Comment:
Move this to the end of the class.
https://google.github.io/styleguide/cppguide.html#Declaration_Order